### PR TITLE
Rocksdb-based metadata storage

### PR DESCRIPTION
**Pull request by cigarl**
In scenarios with too many timeseries, the memory footprint of metadata would become a bottleneck.  Therefore, we propose a scheme to store metadata based on Rocksdb.

Storing metadata in Rocksdb can significantly reduce memory usage.  And we can use appropriate cache size to improve query hit ratio.this reduces the number of _full gc_ and improve service availability.

The `meta_data_manager` configuration item in `iotdb-engine.properties` allows you to select the memory scheme you want to use.  

For example, you can configure it as MEMORY_MANAGER or do no nothing to use the old memory scheme.If you want to use rocksdb to store metadata, you need to change the configuration item to ROCKSDB_MANAGER such as `meta_data_manager=ROCKSDB_MANAGER`.

It's a work in progress, and we'd love to see people testing it and working on it together.  
: )

To make things focus, the PR only includes core implementation of the solution. We'll have other PRs to do things like docs, unit test improvement and etc. 

Design doc: [基于RocksDB的元数据存储 .pptx](https://github.com/apache/iotdb/files/8347564/RocksDB.pptx)

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/MRocksDBManager.java [hunk 1]
```diff
@@ -0,0 +1,2551 @@
...
+  public static int MAX_PATH_DEPTH = 10;
+
+  private static final long MAX_LOCK_WAIT_TIME = 50;
+
+  private RocksDBReadWriteHandler readWriteHandler;
+
+  private final Map<String, ReentrantLock> locksPool =
+      new MapMaker().weakValues().initialCapacity(10000).makeMap();
+
+  private volatile Map<String, Boolean> storageGroupDeletingFlagMap = new ConcurrentHashMap<>();
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/MRocksDBManager.java:189]: The ConcurrentHashMap implementation ensures visibility and does not need to use volatile.

File: server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/MetaDataTransfer.java [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+  MetaDataTransfer() throws MetadataException {
+    rocksDBManager = new MRocksDBManager();
+  }
+
+  public static void main(String[] args) {
+    try {
+      MetaDataTransfer transfer = new MetaDataTransfer();
+      transfer.doTransfer();
+    } catch (MetadataException | IOException | ExecutionException | InterruptedException e) {
+      e.printStackTrace();
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/MetaDataTransfer.java:98]: please use log print ...

File: server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/MetaDataTransfer.java [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+          break;
+        case SET_STORAGE_GROUP:
+        case TTL:
+        case CHANGE_ALIAS:
+        case DELETE_TIMESERIES:
+          executeBufferedOperation(nonCollisionCollections);
+          try {
+            rocksDBManager.operation(plan);
+          } catch (IOException e) {
+            rocksDBManager.operation(plan);
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/MetaDataTransfer.java:197]: add a retry log.

File: server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/MetaDataTransfer.java [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+      for (PhysicalPlan retryPlan : retryPlans) {
+        try {
+          rocksDBManager.operation(retryPlan);
+        } catch (IOException e) {
+          persistFailedLog(retryPlan);
+        } catch (MetadataException e) {
+          logger.error("Execute plan failed: {}", retryPlan.toString(), e);
+        } catch (Exception e) {
+          persistFailedLog(retryPlan);
+        }
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/MetaDataTransfer.java:237]: will IOException add Exception merged

File: server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/RocksDBReadWriteHandler.java [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+    byte[] prefixKey = prefix.getBytes();
+    RocksIterator iterator = rocksDB.newIterator();
+    for (iterator.seek(prefixKey); iterator.isValid(); iterator.next()) {
+      String key = new String(iterator.key());
+      if (!key.startsWith(prefix)) {
+        break;
+      }
+      result.add(key);
+    }
+    //    System.out.println(prefix + " " + result.size());
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/metadata/rocksdb/RocksDBReadWriteHandler.java:337]: delete log

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 6]
```diff
@@ -187,41 +187,11 @@
...
-import static org.apache.iotdb.db.conf.IoTDBConstant.COLUMN_COUNT;
-import static org.apache.iotdb.db.conf.IoTDBConstant.COLUMN_DEVICES;
-import static org.apache.iotdb.db.conf.IoTDBConstant.COLUMN_FUNCTION_CLASS;
-import static org.apache.iotdb.db.conf.IoTDBConstant.COLUMN_FUNCTION_NAME;
-import static org.apache.iotdb.db.conf.IoTDBConstant.COLUMN_FUNCTION_TYPE;
-import static org.apache.iotdb.db.conf.IoTDBConstant.COLUMN_ITEM;
-import static org.apache.iotdb.db.conf.IoTDBConstant.COLUMN_LOCK_INFO;
-import static org.apache.iotdb.db.conf.IoTDBConstant.COLUMN_PRIVILEGE;
-import static org.apache.iotdb.db.conf.IoTDBConstant.COLUMN_ROLE;
+import static org.apache.iotdb.db.conf.IoTDBConstant.*;
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:190]: import *

File: server/src/assembly/resources/conf/iotdb-engine.properties [hunk 7]
```diff
@@ -505,6 +505,14 @@ timestamp_precision=ms
 # Datatype: int
 # query_timeout_threshold=60000
 
+####################
```
Comment (SilverNarcissus) [inline server/src/assembly/resources/conf/iotdb-engine.properties:508]: Please also update doc

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+          .setWriteBufferSize(64 * SizeUnit.KB)
+          .setMaxWriteBufferNumber(6)
+          .setMaxBackgroundJobs(10)
+          .setStatistics(new Statistics())
+          .setLogger(rocksDBLogger);
+
+      final Filter bloomFilter = new BloomFilter(64);
+
+      final BlockBasedTableConfig table_options = new BlockBasedTableConfig();
+      Cache cache = new LRUCache(20L * 1024 * 1024 * 1024, 6);
```
Comment (SilverNarcissus) [inline server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java:134]: This size can be a constant.

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+      table_options
+          .setBlockCache(cache)
+          .setFilterPolicy(bloomFilter)
+          .setBlockSizeDeviation(5)
+          .setBlockRestartInterval(10)
+          .setCacheIndexAndFilterBlocks(true)
+          .setBlockCacheCompressed(new LRUCache(10L * 1024 * 1024 * 1024, 6));
```
Comment (SilverNarcissus) [inline server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java:141]: This size can be a constant.

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+        };
+    return keyExistByTypes(levelKey, types);
+  }
+
+  public CheckKeyResult keyExistByTypes(String levelKey, RMNodeType... types)
+      throws RocksDBException {
+    CheckKeyResult result = new CheckKeyResult();
+    try {
+      Arrays.stream(types)
+          //          .parallel()
```
Comment (SilverNarcissus) [inline server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java:274]: Why this line is commented?

File: server/src/test/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/MRocksDBBenchmark.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+  protected static IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+  private RSchemaRegion rocksDBManager;
+
+  public MRocksDBBenchmark(RSchemaRegion rocksDBManager) {
+    this.rocksDBManager = rocksDBManager;
+  }
+
+  public List<RocksDBBenchmarkTask.BenchmarkResult> benchmarkResults = new ArrayList<>();
+
+  //  public void testStorageGroupCreation(List<SetStorageGroupPlan> storageGroups) {
```
Comment (SilverNarcissus) [inline server/src/test/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/MRocksDBBenchmark.java:47]: This class has too many commented code. Please explain the main function of this class and remove these commented code.

**Review by SilverNarcissus — COMMENTED** [review]
Brilliant work! I try to give some comments and could you add more annotation to some key classes and methods? I will review this PR again, thanks a lot~

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+    //              } catch (MetadataException | IOException e) {
+    //                if (!(e instanceof StorageGroupAlreadySetException)
+    //                    && !(e instanceof PathAlreadyExistException)
+    //                    && !(e instanceof AliasAlreadyExistException)) {
+    //                  errorCount.incrementAndGet();
+    //                }
+    //                logger.error(
+    //                    "create storage group {} failed", sgNode.getPartialPath().getFullPath(),
+    // e);
+    //              }
```
Comment (MarcosZyk) [inline server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java:360]: Remove the commented code.

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java [hunk 13]
```diff
@@ -0,0 +1,446 @@
...
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/** Use the class to transfer data from mlog.bin and mtree-*.snapshot to rocksdb based manager */
+public class SchemaDataTransfer {
+
```
Comment (MarcosZyk) [inline server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java:63]: I think this is a tool for transfer schema from mlog to rocksdb. Shall we implement this as a formal tool which support user pass mlog path.

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/REntityMNode.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+  public REntityMNode(String fullPath, byte[] value, RSchemaReadWriteHandler readWriteHandler) {
+    super(fullPath, readWriteHandler);
+    deserialize(value);
+  }
+
+  @Override
+  public boolean addAlias(String alias, IMeasurementMNode child) {
+
+    // todo add alias
+    return true;
```
Comment (MarcosZyk) [inline server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/REntityMNode.java:55]: Hope for support on this feature.

**Review by MarcosZyk — COMMENTED** [review]
Excellent work! Please have a look at the comments and suggestions.

File: server/src/assembly/resources/conf/iotdb-engine.properties (hunk 7 above)
Comment (cigarl) [reply server/src/assembly/resources/conf/iotdb-engine.properties:508]: This PR focuses on the code, and we need subsequent PR to supplement the detailed documentation.

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java (hunk 13 above)
Comment (chengjianyun) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java]: Removed the implementation from PR as it only designed for schema <= 0.13. We'd add a new implementation once schema of 0.14 is ready.

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java (hunk 8 above)
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java]: done, PTAL~

File: server/src/test/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/MRocksDBBenchmark.java (hunk 11 above)
Comment (cigarl) [reply server/src/test/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/MRocksDBBenchmark.java]: done

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java (hunk 9 above)
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java]: done

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java (hunk 10 above)
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/RSchemaReadWriteHandler.java]: Has been cleared

File: server/src/assembly/resources/conf/iotdb-engine.properties [hunk 15]
```diff
@@ -505,6 +505,14 @@ timestamp_precision=ms
...
+### Metadata Configuration
+####################
+
+# Which metadata manager to be used, right now MEMORY_MANAGER and ROCKSDB_MANAGER are supported. Default MEMORY_MANAGER.
+# Datatype: string
+# meta_data_manager=MEMORY_MANAGER
```
[the remark this answers is not in the document]
Comment (cigarl) [reply server/src/assembly/resources/conf/iotdb-engine.properties]: done,PTAL~

File: server/src/main/java/org/apache/iotdb/db/metadata/MetadataManagerType.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+  MEMORY_MANAGER,
+  ROCKSDB_MANAGER;
+
+  public static MetadataManagerType of(String value) {
+    try {
+      return Enum.valueOf(MetadataManagerType.class, value);
+    } catch (Exception e) {
+      return MEMORY_MANAGER;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/MetadataManagerType.java]: done

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/RInternalMNode.java [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+            RSchemaUtils.convertPartialPathToInner(
+                childName, childNameMaxLevel, RMNodeType.INTERNAL.getValue());
+        readWriteHandler.updateNode(innerName.getBytes(), new byte[] {});
+      } else if (child instanceof RMeasurementMNode) {
+        String innerName =
+            RSchemaUtils.convertPartialPathToInner(
+                childName, childNameMaxLevel, RMNodeType.MEASUREMENT.getValue());
+        // todo all existing attributes of the measurementNode need to be written
+        readWriteHandler.updateNode(innerName.getBytes(), new byte[] {});
+      }
```
[the remark this answers is not in the document]
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/RInternalMNode.java]: done

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/RInternalMNode.java [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+      newChildNode.getAsEntityMNode().setUseTemplate(oldChildNode.isUseTemplate());
+    }
+
+    newChildNode.setSchemaTemplate(oldChildNode.getSchemaTemplate());
+
+    newChildNode.setParent(this);
+
+    this.deleteChild(oldChildName);
+    this.addChild(newChildNode.getName(), newChildNode);
+  }
```
[the remark this answers is not in the document]
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/RInternalMNode.java]: done

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java (hunk 12 above)
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java]: Has been cleared

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java (hunk 13 above)
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/SchemaDataTransfer.java]: This is based on our own usage scenario.  Therefore, I have removed this class and will consider a later integration into `MetadataUpgrader`

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/REntityMNode.java (hunk 14 above)
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/REntityMNode.java]: Aliases are supported, but do not need to be updated in memory.  The comment here is wrong and has been updated.

File: server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/RMeasurementMNode.java [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+
+  @Override
+  public void setAlias(String alias) {
+    this.alias = alias;
+  }
+
+  @Override
+  public TriggerExecutor getTriggerExecutor() {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (cigarl) [reply server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/rocksdb/mnode/RMeasurementMNode.java:134]: ok

File: server/src/assembly/resources/tools/rocksdb-transfer.sh [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+
+echo ---------------------
+echo Use the script the transfer metadata managed by default(pure memory) SchemaEngine to RocksDB based SchemaEngine
```
[the remark this answers is not in the document]
Comment (cigarl) [reply server/src/assembly/resources/tools/rocksdb-transfer.sh]: fixed

File: server/src/assembly/resources/tools/rocksdb-transfer.bat [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+@REM software distributed under the License is distributed on an
+@REM "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+@REM KIND, either express or implied.  See the License for the
+@REM specific language governing permissions and limitations
+@REM under the License.
+@REM
+
+@echo off
+echo ````````````````````````
+echo Use the script the transfer metadata managed by default(pure memory) SchemaEngine to RocksDB based SchemaEngine
```
[the remark this answers is not in the document]
Comment (cigarl) [reply server/src/assembly/resources/tools/rocksdb-transfer.bat]: fixed

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**